src/Arara/Process/IPC/Semaphore.php

```diff
@@ -0,0 +1,132 @@
+<?php
+
+namespace Arara\Process\IPC ;
+
+/**
+ * This class provides Interprocess communication using UNIX Semaphores
+ *
+ * Code written by Nikolay and copied into Arara/Process under the terms of
+ * the MIT license.
+ *
+ * Full credit to Nikolay for his excellent class
+ * 
+ * @license http://opensource.org/licenses/MIT
+ * @author Nikolay Bondarenko http://misterion.ru 
+ */
+class Semaphore
+{
+    /**
+     * @var resource
+     */
+    protected $mutex;
+
+    /**
+     * @var string
+     */
+    protected $file;
+
+    /**
+     * @var bool
+     */
+    protected $isAcquired;
+
+    /**
+     * Creator process pid
+     *
+     * @var int
+     */
+    protected $pid;
+
+    public function __construct($key = null)
+    {
+        $this->pid = getmypid();
+
+        $this->isAcquired = false;
+        if ($key) {
+            $semKey = $key;
+        } else {
+            $this->file = tempnam(sys_get_temp_dir(), 's');
+            $semKey = ftok($this->file, 'a');
+        }
+
+        $this->mutex = sem_get($semKey, 1); //auto_release = 1 by default
+        if (!$this->mutex) {
+            throw new \RuntimeException('Unable to create the semaphore');
+        }
+    }
+
+    public function __destruct()
+    {
+        if ($this->isAcquired()) {
+            $this->release();
+        }
+
+        if ($this->pid === getmypid()) {
+            $this->remove();
+        }
```

Comment:
Oh I see what you mean.  I was only thinking of the releasing, not the removing.  Thanks for the clarification. That is actually a good point that I hadn't considered (ie. using between unrelated processes).  

It might be best to reject this pull request at this stage.  As you have highlighted, there are some issues with this implementation, and I don't really have time to address them.  Since submitting this pull request, I have also modified my algorithm such that I no longer require IPC for my own project.

Thanks for your consideration, and assistance.

Regards,
Damien.